docs/design_proposals/ko-builder.md

```diff
@@ -631,13 +640,17 @@ The steps roughly outlined:
 
     Config options supported, all are optional:
 
-    -   `dependencies`, for Skaffold file watching.
+    -   `fromImage`, to override the default distroless base image
+    -   `dependencies`, for Skaffold file watching
+    -   `dir`, if Go sources are not in the `context` directory
     -   `env`, to support ko CLI users who currently set environment variables
         such as `GOFLAGS` when running ko.
-    -   `fromImage`, to override the default distroless base image
-    -   `labels`
+    -   `labels`, e.g., to
```

Comment:
Aside: does Ko add any labels or other identifying markers that the image was built with ko?

Reply:
Recent versions of ko add the `org.opencontainers.image.base.name` and `org.opencontainers.image.base.digest` labels by default, but the values here don't reference ko.

Are you thinking about how to identify a Go app for debugging purposes?

ko sets the [`KO_DATA_PATH` environment variable](https://github.com/google/ko#static-assets) in the images it builds. This would be a straightforward way to identify a ko image, and it means we can reuse existing Skaffold mechanisms for detecting Go images. I'll add a note to the design proposal for debugging support.

For completeness, some other possible ways to identify a ko-built image from the image config:

1. The `author` field is set to `github.com/google/ko`.

2. The `author` field for the top two layers in the stack will be `ko`.

Here are the manifest and config from the [ko test app](https://github.com/google/ko/tree/main/test) image:

**Manifest**

```json
{
  "schemaVersion": 2,
  "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
  "config": {
    "mediaType": "application/vnd.docker.container.image.v1+json",
    "size": 1014,
    "digest": "sha256:0da11ee446199c9c72cad079fabecd0abfb6981c591b8fd2ff434b1c103bb6a8"
  },
  "layers": [
    {
      "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
      "size": 657718,
      "digest": "sha256:0d7d70899875b098ace120b574f57b39e91415ceaf9b348d6abe168537509f5a"
    },
    {
      "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
      "size": 515,
      "digest": "sha256:799c66aaaf58bb3ff133b65f2106167efda0dd26f0e2dc273c17be8b54790ad7"
    },
    {
      "mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
      "size": 1287324,
      "digest": "sha256:c90b99b518acadaf6340a98a9a3d0af78400f5be7a2a4607255fd31c1f5dab87"
    }
  ],
  "annotations": {
    "org.opencontainers.image.base.digest": "sha256:b871bb2b01374c0a9ed93fdeaa2cdb25b515cd1999170b5ec816ed6c2fd85aca",
    "org.opencontainers.image.base.name": "gcr.io/distroless/static:nonroot"
  }
}
```

**Config**

```json
{
  "architecture": "amd64",
  "author": "github.com/google/ko",
  "created": "1970-01-01T00:00:00Z",
  "history": [
    {
      "author": "Bazel",
      "created": "1970-01-01T00:00:00Z",
      "created_by": "bazel build ..."
    },
    {
      "author": "ko",
      "created": "0001-01-01T00:00:00Z",
      "created_by": "ko publish ko://github.com/google/ko/test",
      "comment": "kodata contents, at $KO_DATA_PATH"
    },
    {
      "author": "ko",
      "created": "0001-01-01T00:00:00Z",
      "created_by": "ko publish ko://github.com/google/ko/test",
      "comment": "go build output, at /ko-app/test"
    }
  ],
  "os": "linux",
  "rootfs": {
    "type": "layers",
    "diff_ids": [
      "sha256:07363fa8421000ad294c2881d17b0535aabdd17ced2a874eb354a9d8514d3d59",
      "sha256:8b4a145477e589ba1218ddc623a80e9774773cbc9c5ffd9e239e60dbb550728b",
      "sha256:3ff3a555f8164bf4c69a49c9623cedecb079e775c3e3ebc68753d6a7400f737a"
    ]
  },
  "config": {
    "Entrypoint": [
      "/ko-app/test"
    ],
    "Env": [
      "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/ko-app/test",
      "SSL_CERT_FILE=/etc/ssl/certs/ca-certificates.crt",
      "KO_DATA_PATH=/var/run/ko"
    ],
    "User": "65532",
    "WorkingDir": "/home/nonroot"
  }
}
```

Reply:
`KO_DATA_PATH` will do for now; the author field seems great too.